File: core/src/main/java/com/vzome/core/commands/CommandCentroid.java


package com.vzome.core.commands;

import java.util.ArrayList;
import java.util.List;

import com.vzome.core.construction.CentroidPoint;
import com.vzome.core.construction.Construction;
import com.vzome.core.construction.ConstructionChanges;
import com.vzome.core.construction.ConstructionList;
import com.vzome.core.construction.Point;

/**
 * @author Scott Vorthmann
 */
public class CommandCentroid extends AbstractCommand
{
    private static final Object[][] PARAM_SIGNATURE =
        new Object[][]{ { GENERIC_PARAM_NAME, Point.class } };

    private static final Object[][] ATTR_SIGNATURE = new Object[][]{};

    @Override
    public Object[][] getParameterSignature()
    {
        return PARAM_SIGNATURE;
    }

    @Override
    public Object[][] getAttributeSignature()
    {
        return ATTR_SIGNATURE;
    }
    
    @Override
    public ConstructionList apply( ConstructionList parameters, AttributeMap attrs, ConstructionChanges effects ) throws Failure
    {
        ConstructionList result = new ConstructionList();
        if ( parameters == null || parameters .size() == 0 )
            throw new Failure( "Select two or more balls to compute their centroid." );
        final Construction[] params = parameters .getConstructions();
        
        List<Point> verticesList = new ArrayList<>();
        for (Construction param : params) {
            if (param instanceof Point) {
                verticesList.add((Point) param);
            }
        }
        // Checking if verticesList .size() < 2 causes old files to fail to load
        // but failing to check if verticesList .isEmpty() throws ArrayIndexOutOfBoundsException.
        // Allowing a single ball to be its own centroid will let older files load,
        // but will generate an appropriate warning if any struts or panels are selected, but no balls.
        if ( verticesList .isEmpty() )
            throw new Failure( "Select two or more balls to compute their centroid." );
        Point[] points = new Point[0];
        CentroidPoint centroid = new CentroidPoint( verticesList .toArray( points ) );
        effects .constructionAdded( centroid );
        result .addConstruction( centroid );
        return result;
    }
}

File: online/src/worker/legacy/ts/com/vzome/core/commands/CommandCentroid.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { AbstractCommand } from "./AbstractCommand.js";
import { AttributeMap } from "./AttributeMap.js";
import { Command } from "./Command.js";
import { CentroidPoint } from "../construction/CentroidPoint.js";
import { Construction } from "../construction/Construction.js";
import { ConstructionChanges } from "../construction/ConstructionChanges.js";
import { ConstructionList } from "../construction/ConstructionList.js";
import { Point } from "../construction/Point.js";

/**
 * @author Scott Vorthmann
 * @class
 * @extends AbstractCommand
 */
export class CommandCentroid extends AbstractCommand {
    static PARAM_SIGNATURE: any[][]; public static PARAM_SIGNATURE_$LI$(): any[][] { if (CommandCentroid.PARAM_SIGNATURE == null) { CommandCentroid.PARAM_SIGNATURE = [[Command.GENERIC_PARAM_NAME, Point]]; }  return CommandCentroid.PARAM_SIGNATURE; }

    static ATTR_SIGNATURE: any[][]; public static ATTR_SIGNATURE_$LI$(): any[][] { if (CommandCentroid.ATTR_SIGNATURE == null) { CommandCentroid.ATTR_SIGNATURE = []; }  return CommandCentroid.ATTR_SIGNATURE; }

    /**
     * 
     * @return {java.lang.Object[][]}
     */
    public getParameterSignature(): any[][] {
        return CommandCentroid.PARAM_SIGNATURE_$LI$();
    }

    /**
     * 
     * @return {java.lang.Object[][]}
     */
    public getAttributeSignature(): any[][] {
        return CommandCentroid.ATTR_SIGNATURE_$LI$();
    }

    /**
     * 
     * @param {ConstructionList} parameters
     * @param {AttributeMap} attrs
     * @param {*} effects
     * @return {ConstructionList}
     */
    public apply(parameters: ConstructionList, attrs: AttributeMap, effects: ConstructionChanges): ConstructionList {
        const result: ConstructionList = new ConstructionList();
        if (parameters == null || parameters.size() === 0)throw new Command.Failure("Select two or more balls to compute their centroid.");
        const params: Construction[] = parameters.getConstructions();
        const verticesList: java.util.List<Point> = <any>(new java.util.ArrayList<any>());
        for(let index = 0; index < params.length; index++) {
            let param = params[index];
            {
                if (param != null && param instanceof <any>Point){
                    verticesList.add(<Point>param);
                }
            }
        }
        if (verticesList.isEmpty())throw new Command.Failure("Select two or more balls to compute their centroid.");
        const points: Point[] = [];
        const centroid: CentroidPoint = new CentroidPoint(verticesList.toArray<any>(points));
        effects['constructionAdded$com_vzome_core_construction_Construction'](centroid);
        result.addConstruction(centroid);
        return result;
    }

    constructor() {
        super();
    }
}
CommandCentroid["__class"] = "com.vzome.core.commands.CommandCentroid";
CommandCentroid["__interfaces"] = ["com.vzome.core.commands.Command"];
